### ISSUE TITLE
Improvements for XDS validation

### ISSUE THREAD

**Issue by Thopap** [oehf/ipf#103]
Hello IPF Team,

i want to share two items that could be improved in the IPF XDS validation:

1. Currently IPF accept XDS folders without having a folder code (which is mandatory from XDS spec)
There is a strange comment in the code which indicate that IPF make this optional due to issues with a testtool. I expect this is already outdated.
org.openehealth.ipf.commons.ihe.xds.core.validate.requests.ObjectContainerValidator.folderSlotValidations
2. Improve error reporting: Currently if  a client sent in a StoredQuery with multiple times the same slot for a value that is not multiple, IPF seems to pick up only one of these slots and ignore the rest. The semantic of sending the same slot twice or more should be interpreted as an "AND", but his make no sense if the attribute is a single one. E.g. FindDocuments Query Support the XDSDocumentEntryStatus, in case a client sent the following, IPF just pick the fist, but such a definition should cause an error.

```xml
<rim:Slot name="$XDSDocumentEntryStatus">
    <rim:ValueList>
        <rim:Value>('urn:oasis:names:tc:ebxml-regrep:StatusType:Approved')</rim:Value>
    </rim:ValueList>
</rim:Slot>
<rim:Slot name="$XDSDocumentEntryStatus">
    <rim:ValueList>
        <rim:Value>('urn:oasis:names:tc:ebxml-regrep:StatusType:Deprecated')</rim:Value>
    </rim:ValueList>
</rim:Slot>
```

To summarize: The validation should validate that only certain slots (e.g. $XDSDocumentEntryEventCodeList) supports a "AND" semantic, all others (e.g. $XDSDocumentEntryStatus) should return an error if defined multiple times.

Regards,
Thomas

**Comment by unixoid** [comment]
@Thopap, please elaborate on the folder code.  Because ITI TF-3, section 4.2.3.4.2 "Folder.codeList", line 2510 states: "Shall have zero or more values."

**Comment by Thopap** [comment]
Figure 4.2.1.3-2: Folder Metadata Attributes points on 1...*

I have opened a discussion in the mailinglist, the current answer from Bill bring up another aspect: https://groups.google.com/forum/#!topic/ihe-xds-implementors/xTrkwOeq0UY

**Comment by unixoid** [comment]
Thank you for the update.  Please let me know when all questions will be clarified as I do not monitor the XDS mailing list.

**Comment by Thopap** [comment]
The disucssion of the mailinglist seems to be finished.

To summarize the issue, I would say that the following definition in the XDS spec make clear that for XDS.b transaction the attribute is mandatory and is optional for XDR + XDM:

> For this question I consider Table 4.3.1.1-3 ("Sending Actor Metadata Attribute Optionality") authoritative. The table 4.3.1.1-3 clearly states that codeList is required for XDS and required if known for XDM and XDR. This shows that the coding section is right - codeList may not have a value - in XDM and XDR. In XDS it must have a value. I would expect MU not to change such a basic guarantee, i.e. after an update, the codeList must still have at least one value.

Is it possible for IPF to separate these concerns (XDM/XDR vs XDS) in the validation?

**Comment by unixoid** [comment]
The class `XdsCamelValidators` now provides three different validators:
* `iti41RequestValidator` -- limited metadata not allowed,
* `iti41XdrRequestValidator` -- limited metadata allowed,
* `iti41XdmRequestValidator` -- limited metadata required.

@Thopap, please test and les us know whether the issue can be closed.

**Comment by Thopap** [comment]
Thank you, folder code validation is now working as expected, you can close this issue.
